### ISSUE TITLE
[Bug] Inconsistent shape with model criterion `ei`

### ISSUE THREAD

**Issue by eddiebergman** [automl/ifBO#16]
Heyo, in NePS, when trying out EI instead of PI, I got an issue where the shape output by `ei` was actually just duplicated vstacked columns, where as for `pi` and `ucb`, it was a single column output as expected.

Here's the repro. Sorry if it's not the most concise, as it's basically extracting out some sample data I used as well as pulling out some internal logic.

Here's the values and shapes for `pi`, `ei` and `ucb`:

```python
tensor([5.9217e-04, 8.1137e-06], grad_fn=<SumBackward1>)
torch.Size([2])
tensor([[4.5409e-05, 2.9934e-07],
        [4.5409e-05, 2.9934e-07]], grad_fn=<ViewBackward0>)
torch.Size([2, 2])
tensor([0.6915, 0.6831], grad_fn=<AddBackward0>)
torch.Size([2])
```

Repro script:

```python
from __future__ import annotations

import torch

x = torch.tensor(
    [
        [1.0000, 0.0000, 0.1036, 1.0000, 0.0312, 0.1230],
        [2.0000, 0.0000, 0.1658, 0.0000, 0.4688, 0.1522],
        [3.0000, 0.0000, 0.3707, 0.2500, 0.6250, 0.8594],
        [4.0000, 0.0000, 0.4301, 0.7500, 0.8750, 0.8967],
        [5.0000, 0.0000, 0.0579, 0.5000, 0.2188, 0.0401],
        [5.0000, 0.1000, 0.0579, 0.5000, 0.2188, 0.0401],
        [6.0000, 0.0000, 0.1061, 0.5000, 0.1719, 0.0600],
        [6.0000, 0.1000, 0.1061, 0.5000, 0.1719, 0.0600],
        [6.0000, 0.2000, 0.1061, 0.5000, 0.1719, 0.0600],
        [6.0000, 0.3000, 0.1061, 0.5000, 0.1719, 0.0600],
        [7.0000, 0.0000, 0.1435, 0.5000, 0.2031, 0.0548],
        [7.0000, 0.1000, 0.1435, 0.5000, 0.2031, 0.0548],
        [8.0000, 0.0000, 0.0847, 0.2500, 0.2344, 0.1516],
        [8.0000, 0.1000, 0.0847, 0.2500, 0.2344, 0.1516],
        [9.0000, 0.0000, 0.1534, 0.5000, 0.2031, 0.1576],
        [9.0000, 0.1000, 0.1534, 0.5000, 0.2031, 0.1576],
        [10.0000, 0.0000, 0.1327, 0.2500, 0.2188, 0.1108],
        [10.0000, 0.1000, 0.1327, 0.2500, 0.2188, 0.1108],
    ],
    dtype=torch.float32,
)
y = torch.tensor(
    [
        0.5,
        0.5,
        0.5,
        0.5,
        0.5,
        0.6,
        0.5,
        0.6,
        0.7,
        0.8,
        0.5,
        0.6,
        0.5,
        0.6,
        0.5,
        0.6,
        0.5,
        0.6,
    ],
    dtype=torch.float32,
)
acq = torch.tensor(
    [
        [9.0000, 0.2000, 0.1534, 0.5000, 0.2031, 0.1576],
        [10.0000, 0.2000, 0.1327, 0.2500, 0.2188, 0.1108],
    ],
    dtype=torch.float32,
)
y_best = torch.full(size=(len(acq),), fill_value=y.max().item(), dtype=torch.float32)

import ifbo

ftpfn = ifbo.FTPFN()


def _cast_tensor_shapes(x: torch.Tensor) -> torch.Tensor:
    if len(x.shape) == 3 and x.shape[1] == 1:
        return x
    if len(x.shape) == 2:
        return x.reshape(x.shape[0], 1, x.shape[1])
    if len(x.shape) == 1:
        return x.reshape(x.shape[0], 1)
    raise ValueError(f"Shape not recognized: {x.shape}")


logits = ftpfn.model(
    _cast_tensor_shapes(x), _cast_tensor_shapes(y), _cast_tensor_shapes(acq)
)

pis = ftpfn.model.criterion.pi(
    logits.squeeze(),
    best_f=y_best.unsqueeze(1),
)
print(pis)  # noqa: T201
print(pis.shape)  # noqa: T201

eis = ftpfn.model.criterion.ei(
    logits.squeeze(),
    best_f=y_best.unsqueeze(1),
)
print(eis)  # noqa: T201
print(eis.shape)  # noqa: T201

ucbs = ftpfn.model.criterion.ucb(
    logits.squeeze(),
    best_f=None,
    rest_prob=(1 - 0.682) / 2,
    maximize=True,
)
print(ucbs)  # noqa: T201
print(ucbs.shape)  # noqa: T201

```

**Comment by eddiebergman** [comment]
Found some more interesting info about EI. In NePS, we explicitly convert everything to maximization for FTPFN and we have this `y_best=1 - y.max()`, which seemed off. I seemed to get it consistenly minimizing for PI but no matter what variation I do, whether I set `y` to be minimization or maximization, and whether I do this `1-` part or not, it seems like the objective always tends towards the worst value.

Not sure what to do with this info but somethings up with EI

**Comment by eddiebergman** [comment]
Alright, I found the problem for both cases! The issue was that within NePS, we were essentially filling up a tensor for `best_f` with `torch.full(shape, y_best)` for all three acquisitions.

After I got rid of some unnessary operations in NePS and just passed a single float through, it seems both EI returns the correct shape and begins to optimize properly.

tldr; `pi` handles a `best_f: torch.Tensor | float` the same while `ei` handles `best_f: torch.Tensor` differently from `ei: float`, where it seems `ei` with `best_f: torch.Tensor` is doing something incorrectly.
